README.md
# myfatoorah_flutter

In order to simplify the integration of your application with MyFatoorah payment platforms, we have developed 
a cutting-edge plugin that works smoothly with your application and provide you with a simple way to embed our payment 
functions within your application. 

The plugin will save your efforts and time instead of integrating with our API using normal API calls, and will allow
you to have the setup ready in a quick, modern and secured way.


## Prerequisites

In order to have the plugin integration working on live environment, please refer to the section 
[Prerequisites](https://myfatoorah.readme.io/v2.0/docs/prerequisites-2) 
and read it for more details

# Integration 


## Installation

##### 1. Add MyFatoorah plugin to your pubspec.yaml file.

    dependencies:
        myfatoorah_flutter: ^2.1.20    
	  
##### 2. Install the plugin by running the following command.

    $ flutter pub get

## Usage
Inside your Dart code do the following:

##### 1. To start using MyFatoorah plugin, import it into your Flutter app. 
 
    import 'package:myfatoorah_flutter/myfatoorah_flutter.dart';

#####  2. Initiate MyFatoorah Plugin inside initState().

    MFSDK.init(<Put API URL here>, MFCountry.KUWAIT, MFEnvironment.TEST);
    
* You can get the `API URL` for testing from [here](https://myfatoorah.readme.io/docs/test-token)	
* Once your testing is finished, simply replace the `API URL` and `environment` with the live information, click
[here](https://myfatoorah.readme.io/docs/live-token) for more information.   
     
#####  3. (Optional).

* Use the following lines if you want to set up the properties of AppBar.

        MFSDK.setUpAppBar(
            title: "MyFatoorah Payment",
            titleColor: Colors.white,  // Color(0xFFFFFFFF)
            backgroundColor: Colors.black, // Color(0xFF000000)
            isShowAppBar: true); // For Android platform only

* And use this line, if you want to hide the AppBar. Note, if the platform is iOS, this line will not affected

        MFSDK.setUpAppBar(isShowAppBar: false);
	

### Initiate / Execute Payment


* Initiate Payment: this step will simply return you all available Payment Methods for the account with the actual
 charge that the customer will pay on the gateway.
  
        var request = new MFInitiatePaymentRequest(0.100, MFCurrencyISO.KUWAIT_KWD);
    
        MFSDK.initiatePayment(request, MFAPILanguage.EN,
                (MFResult<MFInitiatePaymentResponse> result) => {
    
              if(result.isSuccess()) {
                print(result.response?.toJson().toString())
              }
              else {
                print(result.error?.message)
              }
            });

  
* Execute Payment: once the payment has been initiated, this step will do execute the actual transaction creation at 
MyFatoorah platform and will return to your application the URL to redirect your customer to make the payment.
  
        // The value 1 is the paymentMethodId of KNET payment method.
        // You should call the "initiatePayment" API to can get this id and the ids of all other payment methods
        String paymentMethod = 1;
    
        var request = new MFExecutePaymentRequest(paymentMethod, 0.100);
    
        MFSDK.executePayment(context, request, MFAPILanguage.EN,
                (String invoiceId, MFResult<MFPaymentStatusResponse> result) => {
    
              if(result.isSuccess()) {
                print(result.response?.toJson().toString())
              }
              else {
                print(result.error?.message)
              }
            });
  
* As a good practice, you don't have to call the Initiate Payment function every time you need to execute payment, but
 you have to call it at least once to save the PaymentMethodId that you will need to call Execute Payment
  
### Direct Payment / Tokenization

You have to know the following steps to understand how it works:
  
* Get the payment method that allows Direct Payment by calling initiatePayment to get paymentMethodId
* Collect card info from user MFCardInfo(cardNumber: "51234500000000081", cardExpiryMonth: "05", cardExpiryYear: "21", cardSecurityCode: "100", saveToken: false)
* If you want to save your credit card info and get a token for next payment you have to set saveToken: true and you will get the token in the response read more in Tokenization
* If you want to execute a payment through a saved token you have use 

      MFCardInfo(cardToken: "put your token here")
      
* Now you are ready to execute the payment, please check the following sample code.
      
  
    // The value 2 is the paymentMethodId of Visa/Master payment method.
    // You should call the "initiatePayment" API to can get this id and the ids of all other payment methods
    String paymentMethod = 2;

    var request = new MFExecutePaymentRequest(paymentMethod, 0.100);
	
	// var mfCardInfo = new MFCardInfo(cardToken: "Put your token here");

    var mfCardInfo = new MFCardInfo("2223000000000007", "05", "21", "100",
        bypass3DS: false, saveToken: true);

    MFSDK.executeDirectPayment(context, request, mfCardInfo, MFAPILanguage.EN,
            (String invoiceId, MFResult<MFDirectPaymentResponse> result) => {

          if(result.isSuccess()) {
            print(result.response?.toJson().toString())
          }
          else {
            print(result.error?.message)
          }
        });
		
		
		
### Send Payment (Offline)
This will allow you to generate a payment link that can be sent by any channel we support and collect it once it's 
paid by your customer

    var request = MFSendPaymentRequest(invoiceValue: 0.100, customerName: "Customer name",
		notificationOption: MFNotificationOption.LINK);

    MFSDK.sendPayment(context, MFAPILanguage.EN, request, 
            (MFResult<MFSendPaymentResponse> result) => {
      
      if(result.isSuccess()) {
        print(result.response?.toJson().toString())
      }
      else {
        print(result.error?.message)
      }
    });
	
	
	
### Payment Enquiry
This will enable your application to get the full details about a certain invoice / payment
    
    var request = MFPaymentStatusRequest(invoiceId: "12345");

    MFSDK.getPaymentStatus(MFAPILanguage.EN, request,
            (MFResult<MFPaymentStatusResponse> result) => {

          if(result.isSuccess()) {
            print(result.response?.toJson().toString())
          }
          else {
            print(result.error?.message)
          }
        });

  
### Embedded Payment (new)
#### Introduction
If you would like your customer to complete the payment on your checkout page without being redirected to another page (except for the `3D Secure` page) and the `PCI DSS` compliance is a blocker, MyFatoorah embedded payment feature is your optimum solution.

#### Usage

##### Step 1:

Create instance of `MFPaymentCardView` and add it to your `build()` function like the following: 

      @override
      Widget build(BuildContext context) {
        return createPaymentCardView();
      }
    
      createPaymentCardView() {
        if(mfPaymentCardView == null)
            mfPaymentCardView = MFPaymentCardView();
        return mfPaymentCardView;
      }
      
###### Note: you could custom a lot of properties of the payment card view like the following:

    mfPaymentCardView = MFPaymentCardView(
          inputColor: Colors.red,
          labelColor: Colors.yellow,
          errorColor: Colors.blue,
          borderColor: Colors.green,
          fontSize: 14,
          borderWidth: 1,
          borderRadius: 10,
          cardHeight: 220,
          cardHolderNameHint: "card holder name hint",
          cardNumberHint: "card number hint",
          expiryDateHint: "expiry date hint",
          cvvHint: "cvv hint",
          showLabels: true,
          cardHolderNameLabel: "card holder name label",
          cardNumberLabel: "card number label",
          expiryDateLabel: "expiry date label",
          cvvLabel: "securtity code label",
        );
      
##### Step 2:

You need to call `initiateSession()` function to create session. You need to do this for each payment separately. `Session` is valid for only one payment. and inside it's success state, call `load()` function and pass it the session response, to load the payment card view on the screen, like the following:
    
    void initiateSession() {
         /**
         * If you want to use saved card option with embedded payment, send the parameter
         * "customerIdentifier" with a unique value for each customer. This value cannot be used
         * for more than one Customer.
         */
        // var request = MFInitiateSessionRequest("12332212");
    
        /**
         * If not, then send null like this.
         */
        MFSDK.initiateSession(null, (MFResult<MFInitiateSessionResponse> result) => {
          if(result.isSuccess()) 
            mfPaymentCardView.load(result.response!,
                onCardBinChanged: (String bin) => {print("Bin: " + bin)})
          else
            print("Response: " + result.error?.toJson().toString());
        });
      }
###### Note: The `initiateSession()` function should called after `MFSDK.init()` function (that we mentioned above).
##### Step 3:

After that, you need to handle your `Pay` button to call the `pay()` function, copy the below code to your pay event handler section:

    var request = MFExecutePaymentRequest.constructor(0.100);
    
    mfPaymentCardView.pay(
        request,
        MFAPILanguage.EN,
            (String invoiceId, MFResult<MFPaymentStatusResponse> result) =>
        {
          if (result.isSuccess())
            {
              setState(() {
                print("invoiceId: " + invoiceId);
                _response = result.response?.toJson().toString();
                print("Response: " + _response!);
              })
            }
          else
            {
              setState(() {
                print("invoiceId: " + invoiceId);
                var errorJson = result.error?.toJson();
                print("Error: " + errorJson.toString());
                _response = errorJson?.toString();
              })
            }
        });

### Apple Pay Embedded Payment (new for iOS only)
#### Introduction
To provide a better user experience to your Apple Pay users, MyFatoorah is providing the Apple Pay embedded payment. Follow these steps:

#### Usage

##### Step 1:

Create instance of `MFApplePayButton` and add it to your `build()` function like the following:

      @override
      Widget build(BuildContext context) {
        return createApplePayButton();
      }

      createApplePayButton() {
        mfApplePayButton = MFApplePayButton();
        return mfApplePayButton;
      }

##### Step 2:

You need to call `initiateSession()` function to create session. You need to do this for each payment separately. `Session` is valid for only one payment. and inside it's success state, call `load()` function and pass it the session response, like the following:

    void initiateSession() {
        MFSDK.initiateSession(null, (MFResult<MFInitiateSessionResponse> result) => {
          if(result.isSuccess()) 
            loadApplePay(result.response!)
          else
            print(result.error?.toJson().toString())
        });
      }

    void loadApplePay(MFInitiateSessionResponse mfInitiateSessionResponse) {
        var request = MFExecutePaymentRequest.constructorForApplyPay(
        0.100, MFCurrencyISO.KUWAIT_KWD);

        mfApplePayButton.load(
        mfInitiateSessionResponse,
        request,
        MFAPILanguage.EN,
            (String invoiceId, MFResult<MFPaymentStatusResponse> result) =>
        {
            if (result.isSuccess())
            {
                setState(() {
                    print("invoiceId: " + invoiceId);
                    _response = result.response?.toJson().toString();
                    print("Response: " + _response!);
                })
            }
            else
            {
                setState(() {
                    print("invoiceId: " + invoiceId);
                    print("Error: " + result.error!.toJson().toString());
                    _response = result.error?.message;
                }
            });
    }

###### Note: The `initiateSession()` function should called after `MFSDK.init()` function (that we mentioned above).



CHANGELOG.md
## 2.1.20 - 18/5/2023
* fix redirection after `ExecutePayment`.
* Fix some bugs and make overall enhancement.

## 2.1.19 - 18/5/2023
* Remove `callBackUrl` and `errorUrl` from `ExecutePaymentRequest`.
* Fix some bugs and make overall enhancement.

## 2.1.18 - 24/1/2023
* Add `loadWithStartLoading` in Apple PayButton to handle loading stuff until payment done.
* Fix some bugs and make overall enhancement.

## 2.1.17 - 3/1/2023
* Improve apple pay
* Fix some bugs and make overall enhancement.

## 2.1.16 - 27/11/2022
* Fix some bugs and make overall enhancement.

## 2.1.15 - 21/11/2022
* Add `onCardBinChanged` callback to the `load()` method of `MFPaymentCardView`.
* Fix some bugs and make overall enhancement.

## 2.1.14 - 7/11/2022
* Fix some bugs and make overall enhancement.
 
* ## 2.1.13 - 28/8/2022
* Fix some bugs and make overall enhancement.

## 2.1.12 - 11/5/2022
* Fix some bugs and make overall enhancement.

## 2.1.11 - 8/5/2022
* Fix some bugs and make overall enhancement.

## 2.1.10 - 27/4/2022
* Add `onInvoiceCreated` callback to `executePayment` to return you the `invoiceId` once created.
* Fix some bugs and make overall enhancement.

## 2.1.9 - 24/4/2022
* Add feature of saved cards with embedded payment.
* Fix some bugs and make overall enhancement.

## 2.1.8 - 13/4/2022
* Fix some bugs and make overall enhancement.

## 2.1.7 - 12/4/2022
* Fix some bugs and make overall enhancement.

## 2.1.6 - 10/4/2022
* Add feature of pay with Apple Pay.
* Support save card option with embedded payment to enable you to save your customer's card details to make future payments easier.
* Fix some bugs and make overall enhancement.

## 2.1.5 - 25/1/2022
* Deprecated the `executeRecurringDirectPayment()` method and use instead `RecurringModel` in `MFExecutePaymentRequest` using `executePayment()` or `executeDirectPayment()` method.
* Fix some bugs and make overall enhancement.

## 2.1.4 - 17/1/2022
* Fix some bugs and make overall enhancement.

## 2.1.3 - 29/12/2021
* Fix some bugs and make overall enhancement.

## 2.1.2 - 12/12/2021
* Fix some bugs and make overall enhancement.

## 2.1.0 - 18/10/2021
* Support embedded payment feature (payment card view).
* Fix some bugs and make overall enhancement.

## 2.0.3 - 14/9/2021
* Fix some bugs and make overall enhancement.

## 2.0.2 - 12/9/2021
* Fix some bugs and make overall enhancement.

## 2.0.1 - 15/8/2021
* Fix some bugs and make overall enhancement.

## 2.0.0 - 7/7/2021
* Migrate to null safety.

## 1.0.20 - 14/6/2021
* Fix some bugs and make overall enhancement.

## 1.0.19 - 8/6/2021
* Fix some bugs and make overall enhancement.

## 1.0.18 - 5/5/2021
* Upgrade `device_info` plugin to version `2.0.0`

## 1.0.17 - 13/4/2021
* Upgrade `http` plugin to version `0.13.1`

## 1.0.16 - 24/3/2021
* Deprecated `executeDirectPaymentWithRecurring()` method which takes integer 
value `intervalDays` as a parameter. Used instead `executeRecurringDirectPayment()` method
which takes `MFRecurringType` which is one of four different values 
`daily`, `weekly`, `monthly`, or `custom(days)`.

## 1.0.15 - 17/2/2021
* We deprecated 'supplierCode' and 'supplierValue' and used 'suppliers' list instead.
 Inside the following APIs: 
    - sendPayment
    - executePayment
    - executeDirectPayment
    - executeDirectPaymentWithRecurring
* Fix some bugs and make overall enhancement.

## 1.0.14 - 8/2/2021
* Fix some bugs and make overall enhancement.

## 1.0.13 - 27/1/2021
* Fix some bugs and make overall enhancement.

## 1.0.12 - 11/11/2020
* Fix some bugs and make overall enhancement.

## 1.0.11 - 2/11/2020
* Add CardHolderName field to card info in direct payment.

## 1.0.10 - 19/10/2020
* Fix some bugs and make overall enhancement.

## 1.0.9 - 24/9/2020
* Fix some bugs and make overall enhancement.

## 1.0.8 - 24/9/2020
* Fix some bugs and make overall enhancement.

## 1.0.7 - 16/6/2020
* Fix some bugs and make overall enhancement.

## 1.0.6 - 31/5/2020
* Support Apple Pay for iPhone devices starting from iOS 13.

## 1.0.5 - 19/5/2020
* Fix some bugs and make overall enhancement.

## 1.0.4 - 13/5/2020
* Fix some bugs and make overall enhancement.

## 1.0.3 - 26/3/2020
* Fix some bugs and make overall enhancement.

## 1.0.2 - 25/3/2020

* Add cancelRecurringPayment feature.
* Add cancelToken feature.
* Updating ReadMe and fix some bugs and enhancement.

## 1.0.1 - 24/3/2020

* Adds getPaymentStatus feature.
* Fux bugs and enhancement.

## 1.0.0 - 24/3/2020

* Adds sendPayment feature.
* Adds initiatePayment feature.
* Adds executeRegularPayment feature.
* Adds executeDirectPayment feature.
* Adds executeDirectPaymentWithRecurring feature.

## 0.0.1 - 22/3/2020

* First Beta version of MyFatoorah Flutter Plugin.


example/lib/main.dart
// ignore_for_file: deprecated_member_use

import 'package:flutter/material.dart';
import 'package:myfatoorah_flutter/embeddedapplepay/MFApplePayButton.dart';
import 'package:myfatoorah_flutter/myfatoorah_flutter.dart';
import 'package:myfatoorah_flutter/utils/MFCountry.dart';
import 'package:myfatoorah_flutter/utils/MFEnvironment.dart';
import 'dart:io' show Platform;

/*
TODO: The following API token key for testing only, so that when you go live
      don't forget to replace the following key with the live key provided
      by MyFatoorah.
*/

// You can get the API Token Key from here:
// https://myfatoorah.readme.io/docs/test-token
final String mAPIKey =
    "rLtt6JWvbUHDDhsZnfpAhpYk4dxYDQkbcPTyGaKp2TYqQgG7FGZ5Th_WD53Oq8Ebz6A53njUoo1w3pjU1D4vs_ZMqFiz_j0urb_BH9Oq9VZoKFoJEDAbRZepGcQanImyYrry7Kt6MnMdgfG5jn4HngWoRdKduNNyP4kzcp3mRv7x00ahkm9LAK7ZRieg7k1PDAnBIOG3EyVSJ5kK4WLMvYr7sCwHbHcu4A5WwelxYK0GMJy37bNAarSJDFQsJ2ZvJjvMDmfWwDVFEVe_5tOomfVNt6bOg9mexbGjMrnHBnKnZR1vQbBtQieDlQepzTZMuQrSuKn-t5XZM7V6fCW7oP-uXGX-sMOajeX65JOf6XVpk29DP6ro8WTAflCDANC193yof8-f5_EYY-3hXhJj7RBXmizDpneEQDSaSz5sFk0sV5qPcARJ9zGG73vuGFyenjPPmtDtXtpx35A-BVcOSBYVIWe9kndG3nclfefjKEuZ3m4jL9Gg1h2JBvmXSMYiZtp9MR5I6pvbvylU_PP5xJFSjVTIz7IQSjcVGO41npnwIxRXNRxFOdIUHn0tjQ-7LwvEcTXyPsHXcMD8WtgBh-wxR8aKX7WPSsT1O8d8reb2aR7K3rkV3K82K_0OgawImEpwSvp9MNKynEAJQS6ZHe_J_l77652xwPNxMRTMASk1ZsJL";

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'MyFatoorah Plugin',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'MyFatoorah Plugin Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String _response = '';
  String _loading = "Loading...";
  var sessionIdValue = "";
  late MFPaymentCardView mfPaymentCardView;
  late MFApplePayButton mfApplePayButton;

  @override
  void initState() {
    super.initState();

    if (mAPIKey.isEmpty) {
      setState(() {
        _response =
            "Missing API Token Key.. You can get it from here: https://myfatoorah.readme.io/docs/test-token";
      });
      return;
    }

    // TODO, don't forget to init the MyFatoorah Plugin with the following line
    MFSDK.init(mAPIKey, MFCountry.KUWAIT, MFEnvironment.TEST);
    initiateSession();
    // (Optional) un comment the following lines if you want to set up properties of AppBar.

//    MFSDK.setUpAppBar(
//      title: "MyFatoorah Payment",
//      titleColor: Colors.white,  // Color(0xFFFFFFFF)
//      backgroundColor: Colors.lightBlue, // Color(0xFF000000)
//      isShowAppBar: true); // For Android platform only

    // (Optional) un comment this line, if you want to hide the AppBar.
    // Note, if the platform is iOS, this line will not affected

    MFSDK.setUpAppBar(isShowAppBar: false);
  }

  void initiateSession() {
    /**
     * If you want to use saved card option with embedded payment, send the parameter
     * "customerIdentifier" with a unique value for each customer. This value cannot be used
     * for more than one Customer.
     */
    // var request = MFInitiateSessionRequest("12332212");

    /**
     * If not, then send null like this.
     */
    MFSDK.initiateSession(
        null,
        (MFResult<MFInitiateSessionResponse> result) => {
              if (result.isSuccess())
                {
                  // This for embedded payment view
                  mfPaymentCardView.load(result.response!,
                      onCardBinChanged: (String bin) => {print("Bin: " + bin)}),

                  /// This used for Apple Pay in iOS only.

                  if (Platform.isIOS) loadApplePay(result.response!)
                }
              else
                {
                  setState(() {
                    // print("Response: " + result.error!.toJson().toString());
                    _response = result.error!.message!;
                  })
                }
            });
  }

  void payWithEmbeddedPayment() {
    var request = MFExecutePaymentRequest.constructor(0.100);
    mfPaymentCardView.pay(
        request,
        MFAPILanguage.EN,
        (String invoiceId, MFResult<MFPaymentStatusResponse> result) => {
              if (result.isSuccess())
                {
                  setState(() {
                    print("invoiceId: " + invoiceId);
                    print("Response: " + result.response!.toJson().toString());
                    _response = result.response!.toJson().toString();
                  })
                }
              else
                {
                  setState(() {
                    print("invoiceId: " + invoiceId);
                    print("Error: " + result.error!.toJson().toString());
                    _response = result.error!.message!;
                  })
                }
            });
  }

  /// This used for Apple Pay in iOS only.
  void loadApplePay(MFInitiateSessionResponse mfInitiateSessionResponse) {
    var request = MFExecutePaymentRequest.constructorForApplyPay(
        0.100, MFCurrencyISO.KUWAIT_KWD);
    mfApplePayButton.loadWithStartLoading(
        mfInitiateSessionResponse, request, MFAPILanguage.EN, () {
      setState(() {
        _response = "Loading...";
      });
    },
        (String invoiceId, MFResult<MFPaymentStatusResponse> result) => {
              if (result.isSuccess())
                {
                  setState(() {
                    print("invoiceId: " + invoiceId);
                    print("Response: " + result.response!.toJson().toString());
                    _response = result.response!.toJson().toString();
                  })
                }
              else
                {
                  setState(() {
                    print("invoiceId: " + invoiceId);
                    print("Error: " + result.error!.toJson().toString());
                    _response = result.error!.message!;
                  })
                }
            });
  }

  /*
    Send Payment
   */
  void sendPayment() {
    var request = MFSendPaymentRequest(
        invoiceValue: 0.100,
        customerName: "Customer name",
        notificationOption: MFNotificationOption.LINK);

    MFSDK.sendPayment(
        context,
        MFAPILanguage.EN,
        request,
        (MFResult<MFSendPaymentResponse> result) => {
              if (result.isSuccess())
                {
                  setState(() {
                    print("Response: " + result.response!.toJson().toString());
                    _response = result.response!.toJson().toString();
                  })
                }
              else
                {
                  setState(() {
                    print("Response: " + result.error!.toJson().toString());
                    _response = result.error!.message!;
                  })
                }
            });

    setState(() {
      _response = _loading;
    });
  }

  /*
    Initiate Payment
   */
  void initiatePayment() {
    var request = new MFInitiatePaymentRequest(5.5, MFCurrencyISO.KUWAIT_KWD);

    MFSDK.initiatePayment(
        request,
        MFAPILanguage.EN,
        (MFResult<MFInitiatePaymentResponse> result) => {
              if (result.isSuccess())
                {
                  setState(() {
                    print("Response: " + result.response!.toJson().toString());
                    _response = result.response!.toJson().toString();
                  })
                }
              else
                {
                  setState(() {
                    print("Response: " + result.error!.toJson().toString());
                    _response = result.error!.message!;
                  })
                }
            });

    setState(() {
      _response = _loading;
    });
  }

  /*
    Execute Regular Payment
   */
  void executeRegularPayment() {
    // The value 1 is the paymentMethodId of KNET payment method.
    // You should call the "initiatePayment" API to can get this id and the ids of all other payment methods
    int paymentMethod = 1;

    var request = new MFExecutePaymentRequest(paymentMethod, 0.100);

    MFSDK.executePayment(context, request, MFAPILanguage.EN,
        onInvoiceCreated: (String invoiceId) =>
            {print("invoiceId: " + invoiceId)},
        onPaymentResponse: (String invoiceId,
                MFResult<MFPaymentStatusResponse> result) =>
            {
              if (result.isSuccess())
                {
                  setState(() {
                    print("invoiceId: " + invoiceId);
                    print("Response: " + result.response!.toJson().toString());
                    _response = result.response!.toJson().toString();
                  })
                }
              else
                {
                  setState(() {
                    print("invoiceId: " + invoiceId);
                    print("Response: " + result.error!.toJson().toString());
                    _response = result.error!.message!;
                  })
                }
            });

    setState(() {
      _response = _loading;
    });
  }

  /*
    Execute Direct Payment
   */
  void executeDirectPayment() {
    // The value 9 is the paymentMethodId of Visa/Master payment method.
    // You should call the "initiatePayment" API to can get this id and the
    // ids of all other payment methods
    int paymentMethod = 9;

    var request = new MFExecutePaymentRequest(paymentMethod, 0.100);

//    var mfCardInfo = new MFCardInfo(cardToken: "Put your API token key here");

    var mfCardInfo = new MFCardInfo(
        cardNumber: "5453010000095489",
        expiryMonth: "05",
        expiryYear: "21",
        securityCode: "100",
        cardHolderName: "Set Name",
        bypass3DS: false,
        saveToken: false);

    MFSDK.executeDirectPayment(
        context,
        request,
        mfCardInfo,
        MFAPILanguage.EN,
        (String invoiceId, MFResult<MFDirectPaymentResponse> result) => {
              if (result.isSuccess())
                {
                  setState(() {
                    print("invoiceId: " + invoiceId);
                    print("Response: " + result.response!.toJson().toString());
                    _response = result.response!.toJson().toString();
                  })
                }
              else
                {
                  setState(() {
                    print("invoiceId: " + invoiceId);
                    print("Response: " + result.error!.toJson().toString());
                    _response = result.error!.message!;
                  })
                }
            });

    setState(() {
      _response = _loading;
    });
  }

  /*
    Execute Direct Payment with Recurring
   */
  void executeDirectPaymentWithRecurring() {
    // The value 20 is the paymentMethodId of Visa/Master payment method.
    // You should call the "initiatePayment" API to can get this id and the ids
    // of all other payment methods
    int paymentMethod = 20;

    var request = new MFExecutePaymentRequest(paymentMethod, 100.0);

    var mfCardInfo = new MFCardInfo(
        cardNumber: "5453010000095539",
        expiryMonth: "05",
        expiryYear: "21",
        securityCode: "100",
        cardHolderName: "Set Name",
        bypass3DS: true,
        saveToken: true);

    mfCardInfo.iteration = 12;

    MFSDK.executeRecurringDirectPayment(
        context,
        request,
        mfCardInfo,
        MFRecurringType.monthly,
        MFAPILanguage.EN,
        (String invoiceId, MFResult<MFDirectPaymentResponse> result) => {
              if (result.isSuccess())
                {
                  setState(() {
                    print("Response: " + invoiceId);
                    print("Response: " + result.response!.toJson().toString());
                    _response = result.response!.toJson().toString();
                  })
                }
              else
                {
                  setState(() {
                    print("Response: " + invoiceId);
                    print("Response: " + result.error!.toJson().toString());
                    _response = result.error!.message!;
                  })
                }
            });

    setState(() {
      _response = _loading;
    });
  }

  /*
    Payment Enquiry
   */
  void getPaymentStatus() {
    var request = MFPaymentStatusRequest(
        invoiceId: "1849178"); // 647116 for success 1849178 for fail

    MFSDK.getPaymentStatus(
        MFAPILanguage.EN,
        request,
        (MFResult<MFPaymentStatusResponse> result) => {
              if (result.isSuccess())
                {
                  setState(() {
                    print("Response: " + result.response!.toJson().toString());
                    _response = result.response!.toJson().toString();
                  })
                }
              else
                {
                  setState(() {
                    print("Response: " + result.error!.toJson().toString());
                    _response = result.error!.message!;
                  })
                }
            });

    setState(() {
      _response = _loading;
    });
  }

  /*
    Cancel Token
   */
  void cancelToken() {
    MFSDK.cancelToken(
        "Put your token here",
        MFAPILanguage.EN,
        (MFResult<bool> result) => {
              if (result.isSuccess())
                {
                  setState(() {
                    print("Response: " + result.response.toString());
                    _response = result.response.toString();
                  })
                }
              else
                {
                  setState(() {
                    print("Response: " + result.error!.toJson().toString());
                    _response = result.error!.message!;
                  })
                }
            });

    setState(() {
      _response = _loading;
    });
  }

  /*
    Cancel Recurring Payment
   */
  void cancelRecurringPayment() {
    MFSDK.cancelRecurringPayment(
        "Put RecurringId here",
        MFAPILanguage.EN,
        (MFResult<bool> result) => {
              if (result.isSuccess())
                {
                  setState(() {
                    print("Response: " + result.response.toString());
                    _response = result.response.toString();
                  })
                }
              else
                {
                  setState(() {
                    print("Response: " + result.error!.toJson().toString());
                    _response = result.error!.message!;
                  })
                }
            });

    setState(() {
      _response = _loading;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Column(children: [
          Expanded(
            flex: 1,
            child: SingleChildScrollView(
              child: Column(
//                    mainAxisAlignment: MainAxisAlignment.center,
//                    crossAxisAlignment: CrossAxisAlignment.stretch,
                children: [
                  Padding(
                    padding: EdgeInsets.all(5.0),
                  ),
                  createPaymentCardView(),
                  Padding(
                    padding: EdgeInsets.all(5.0),
                  ),
                  if (Platform.isIOS) createApplePayView(),
                  ElevatedButton(
                    child: Text('Pay (Embedded Payment)'),
                    onPressed: payWithEmbeddedPayment,
                  ),
                  ElevatedButton(
                    child: Text('Send Payment'),
                    onPressed: sendPayment,
                  ),
                  ElevatedButton(
                    child: Text('Initiate Payment'),
                    onPressed: initiatePayment,
                  ),
                  ElevatedButton(
                    child: Text('Execute Regular Payment'),
                    onPressed: executeRegularPayment,
                  ),
                  ElevatedButton(
                    child: Text('Execute Direct Payment'),
                    onPressed: executeDirectPayment,
                  ),
                  ElevatedButton(
                    child: Text('Execute Direct Payment with Recurring'),
                    onPressed: executeDirectPaymentWithRecurring,
                  ),
                  ElevatedButton(
                    child: Text('Cancel Recurring Payment'),
                    onPressed: cancelRecurringPayment,
                  ),
                  ElevatedButton(
                    child: Text('Cancel Token'),
                    onPressed: cancelToken,
                  ),
                  ElevatedButton(
                    child: Text('Get Payment Status'),
                    onPressed: getPaymentStatus,
                  ),
                  Padding(
                    padding:
                        const EdgeInsets.only(top: 10, left: 5.0, right: 5.0),
                    child: Text(
                      _response,
                      style: TextStyle(backgroundColor: Colors.yellowAccent),
                    ),
                  )
                ],
              ),
            ),
          ),
        ]),
      ),
    );
  }

  createPaymentCardView() {
    mfPaymentCardView = MFPaymentCardView(
        // language: MFAPILanguage.EN,
        // inputColor: Colors.red,
        // labelColor: Colors.yellow,
        // errorColor: Colors.blue,
        // borderColor: Colors.green,
        // fontSize: 14,
        // borderWidth: 1,
        // borderRadius: 10,
        // cardHeight: 220,
        // cardHolderNameHint: "card holder name hint",
        // cardNumberHint: "card number hint",
        // expiryDateHint: "expiry date hint",
        // cvvHint: "cvv hint",
        // showLabels: true,
        // cardHolderNameLabel: "card holder name label",
        // cardNumberLabel: "card number label",
        // expiryDateLabel: "expiry date label",
        // cvvLabel: "cvv label",
        );

    return mfPaymentCardView;
  }

  /// This for Apple pay button
  createApplePayView() {
    mfApplePayButton =
        MFApplePayButton(height: 50, radius: 15, buttonText: "Buy with");
    return mfApplePayButton;
  }
}
